pkg/metrics/rate.go

```diff
@@ -35,13 +35,14 @@ func (r *Rate) flush(timestamp float64) ([]*Serie, error) {
 	}
 
 	value, ts := (r.sample-r.previousSample)/(r.timestamp-r.previousTimestamp), r.timestamp
-	r.previousSample, r.previousTimestamp = r.sample, r.timestamp
-	r.sample, r.timestamp = 0., 0.
 
 	if value < 0 {
-		return []*Serie{}, fmt.Errorf("Rate value is negative, discarding it (the underlying counter may have been reset)")
+		return []*Serie{}, fmt.Errorf("Rate value is negative (%f - %f = %f), discarding it (the underlying counter may have been reset)", r.sample, r.previousSample, value)
```

Comment:
if we don't update `r.previousSample` and we've wrapped around a counter, for example, we will continue to have negative rates in subsequent iterations. I believe we want to update those always, even when `value<0`.

Reply:
Makes sense - just pushed an update that assigns some temp variables for the debug output to keep things otherwise working as before.